shopfloor_base/services/service.py

```diff
@@ -0,0 +1,267 @@
+        domain = expression.normalize_domain(self._get_base_search_domain())
+        domain = expression.AND([domain, [("id", "=", _id)]])
+        record = self.env[self._expose_model].search(domain)
+        if not record:
+            raise exceptions.MissingError(
+                _("The record %s %s does not exist") % (self._expose_model, _id)
+            )
+        else:
+            return record
+
+    def _get_base_search_domain(self):
+        return []
+
+    def _convert_one_record(self, record):
+        """To implement in service Components"""
+        return {}
+
+    def _to_json(self, records):
+        res = []
+        for record in records:
+            res.append(self._convert_one_record(record))
+        return res
+
+    def _response(
+        self, base_response=None, data=None, next_state=None, message=None, popup=None
+    ):
+        """Base "envelope" for the responses
+
+        All the keys are optional.
+
+        :param base_response: optional dictionary of values to extend
+        (typically already created by a call to _response())
+        :param data: dictionary of values, when a next_state is provided,
+        the data is enclosed in a key of the same name (to support polymorphism
+        in the schema)
+        :param next_state: string describing the next state that the client
+        application must reach
+        :param message: dictionary for the message to show in the client
+        application (see ``_response_schema`` for the keys)
+        :param popup: dictionary for a popup to show in the client application
+        (see ``_response_schema`` for the keys). The popup is displayed before
+        reaching the next state.
+        """
+        if base_response:
+            response = base_response.copy()
+        else:
+            response = {}
+        if next_state:
+            # data for a state is always enclosed in a key with the name
+            # of the state, so an endpoint can return to different states
+            # that need different data: the schema can be different for
+            # every state this way
+            response.update(
+                {
+                    # ensure we have an empty dict when the state
+                    # does not need any data, so the client does not need
+                    # to check this
+                    "data": {next_state: data or {}},
+                    "next_state": next_state,
+                }
+            )
+
+        elif data:
+            response["data"] = data
+
+        if message:
+            response["message"] = message
+
+        if popup:
+            response["popup"] = popup
+
+        return response
+
+    _requires_header_menu = False
+    _requires_header_profile = False
+
+    def _get_openapi_default_parameters(self):
+        defaults = super()._get_openapi_default_parameters()
+        # Normal users can't read an API key, ignore it using sudo() only
+        # because it's a demo key.
```

Comment:
IMO This code should be put somewhere else but not in the base module.

Reply:
These params are mandatory for the whole app ATM. I won't change in this PR. I can add a TODO to discuss about it later.

Reply:
ok